### ISSUE TITLE
Exec FFMPEG problems only when using hardware decoding?

### ISSUE THREAD

**Issue by MasterEvilAce** [sergey-dryabzhinsky/nginx-rtmp-module#278]
I run an nginx server on a Raspberry Pi3, using the rtmp module. Everything in my standard setup works correctly.

Recently I've been testing transcoding and I've run into an issue that completely stumps me. I've got two applications, "transcode" and "stream". If I stream to the "transcode" app, it'll exec FFMPEG and send two copies of the stream to the "stream" app... one is just a source copy (_hi) and the other is a lower bitrate transcoded copy (_lo). The stream app has HLS enabled.

Transcoding "works" if I allow libx264 to decode and encode the stream, but it's unwatchable due to the raspberry Pi CPU being underpowered and using 99% CPU for this task.
I've been testing using the hardware decoder (h264_mmal) along with the hardware encoder (h264_omx), and the Raspberry pi is able to handle this perfectly fine if I run the FFMPEG command OUTSIDE of nginx. While FFMPEG is in the exec block, it outputs to a log file but it seems to just stop randomly, before outputting anything. Additionally, the rtmp module does not create ANY HLS output.

Here's my manual FFMPEG command that works:
(Obviously here I'm streaming directly to the stream app instead of transcode, just so I can bypass the EXEC line, but it should make no real difference, I suspect, since I can stream to transcode with CPU encoding/decoding and nginx/rtmp module works correctly.)
> ffmpeg -c:v h264_mmal -i rtmp://127.0.0.1/stream/STREAMID -c copy -f flv rtmp://127.0.0.1/stream/STREAMID_hi -c:a copy -c:v h264_omx -b:v 1712k -f flv rtmp://127.0.0.1/stream/STREAMID_lo

RELEVANT NGINX CONFIG
> application transcode {
> 	live on;
> 	exec ffmpeg -c:v h264_mmal -i rtmp://127.0.0.1/$app/$name
> 		 -c copy -f flv rtmp://127.0.0.1/stream/$name_hi
> 		 -c:a copy -c:v h264_omx -b:v 1712k
> 		 -f flv rtmp://127.0.0.1/stream/$name_lo
> 		 -loglevel debug 2>/tmp/ffmpeg.log;
> }
> application stream {
> 	live on;
> 	
> 	hls on;
> 	hls_path /tmp/hls;
> 	hls_type live;
> 	hls_nested off;
> 	hls_continuous off;
> 	hls_cleanup on;
> 
> 	hls_variant _hi NAME=HIGH RESOLUTION=1280x720 BANDWIDTH=3584000;
> 	hls_variant _lo NAME=LOW  RESOLUTION=1280x720 BANDWIDTH=1872000;
> 
> 	hls_fragment         1s;
> 	hls_max_fragment     2s;
> 	hls_playlist_length 30s;
> }

FFMPEG.LOG
This is the FULL log, while FFMPEG is being called by the transcode EXEC line
https://pastebin.com/pRw6eBM8

**Comment by MasterEvilAce** [comment]
Alright... reading through a bunch of threads, just as a test to rule things out... I set nginx.conf to run user as root (instead of www-data) and suddenly the issue is gone and it runs flawlessly.

My question now is... why do the hardware encoder/decoders need extra permissions, and what are they? If I can just give www-data those permissions we should be good.

Any ideas?

**Comment by MasterEvilAce** [comment]
Figured it out. I hope this can help people in the future.

In order to figure out what the error was... I created a new user www-test and added it to the www-data group. I then used this command to run the ffmpeg string as the user www-test
sudo runuser -l  www-test -c '<FFMPEG COMMAND HERE>'
This spit out the error as the very last line which happend to be
* failed to open vchiq instance

Doing more research, all I had to do was add www-data to the video group. 
sudo usermod -a -G video www-data

Then I restarted the machine in order for the changes to take effect. After that, all problems solved. I noticed a lot of people having similar issues in this project as well as the original, and most of them went unsolved (except by changing user to ROOT, which is terrible.)
